### ISSUE TITLE
Override operators in Layer class

### ISSUE THREAD

**Issue by 317070** [Lasagne/Lasagne#735]
Right now, in my model I am using the ElemwiseMergeLayer a lot. Especially since residual networks and skip layers, this has become a very popular approach (compared to when lasagne started).

Could we override the __add__, __radd__, __mul__ and __rmul__ on the Layer class, to create a suitable ElemwiseMergeLayer? This will clean the code for some models considerably. And it would be more intuitive, I reckon.

I have been thinking for a while, but I don't see what would be an alternative operation performed by the operator in this context. I think it is kind of straightforward.

I would implement it myself, and would provide support for adding (and multiplying) constants and numpy arrays. If there is support for this feature from the main developers, of course.

**Comment by f0k** [comment]
With #678, layers will be able to return multiple outputs in form of a list or dictionary. Addition and multiplication could mean different things then:
```
[a, b, c] * 3 -> [a*3, b*3, c*3] or [a, b, c, a, b, c, a, b, c]
[a, b, c] + 3 -> [a+3, b+3, c+3] or [a, b, c, 3]
[a, b, c] + [d, e] -> [a, b, c, d, e]
```
It's also possible that `layer1 + layer2` might be mistaken as concatenating the features.
I can relate to the elegance of writing something like `1*layer1 + 3*layer2`, though.

What do others think?

**Comment by 317070** [comment]
I was not aware of the layers returning lists. I do not know the use case of layers returning multiple tensors, but I reckon detecting them with an `isintstance` and throwing an Exception when __add__ or __mul__ are used on them, will make it hard to have mistakes.

**Comment by f0k** [comment]
> I do not know the use case of layers returning multiple tensors

LSTMs returning their cell state along with their hidden state, for example.

> but I reckon detecting them with an `isintstance` and throwing an Exception

True, that would be a way out. And I think `layer1 + layer2` probably won't be interpreted as concatenation just as `array1 + array2` isn't, which was my other worry. So I think we could try. @benanne, what do you think? Is there anything else addition or multiplication of layers could be confused with?

By the way, for ResNets, if you're not doing projection shortcuts, but zero-padding, it's much more efficient to use a `T.inc_subtensor` layer instead of actually zero-padding and adding. We can't use an operator for that, though. (`+=` doesn't make sense for this since the operators are of different size.)

**Comment by ebenolson** [comment]
While I think it would be amazing if it was possible to compose an abitrary expression and get a `Layer` back, this seems like a pretty radical departure from the current API, and can only support a small subset of what the current `ElemwiseMergeLayer` does.

**Comment by benanne** [comment]
It's difficult for me to judge whether the use cases are important enough to warrant the additional complexity here. Overall I think we should follow "explicit is better than implicit", and for the specific use case of resnets we should encourage people to write helper functions to create parts of the network. But it's true that this is becoming an extremely common use case, so I think I wouldn't be opposed to it in this particular instance.

I do worry that it will serve to confuse newcomers or casual readers of such code though -- '+' can then operate on three different abstraction levels: direct addition (e.g. numpy), addition of Theano expressions, and addition of Lasagne layers. It's easy to confuse any two of these.

**Comment by f0k** [comment]
> if it was possible to compose an abitrary expression and get a `Layer` back

That's basically the philosophy of Blocks. We won't go that far, though, so this could be even more confusing:

> '+' can then operate on three different abstraction levels: direct addition (e.g. numpy), addition of Theano expressions, and addition of Lasagne layers

If we plan to only keep it at `+` and `*`, maybe we shouldn't start it, so we're not pressed into adding `-`, `/`, `**`, and finally make `Layer` a subclass of `theano.Variable` to end the confusion of what's possible.

> It's difficult for me to judge whether the use cases are important enough to warrant the additional complexity here.

@317070, do you have a concrete code example? Would operators really change that much? You could still do `from lasagne.layers import ElemwiseSumLayer as add`, if you wanted. If you're only adding two layers, polish notation is not so much worse than an infix operator, is it?

**Comment by 317070** [comment]
> If we plan to only keep it at `+` and `*`, maybe we shouldn't start it, so we're not pressed into adding `-`, `/`, `**`, and finally make `Layer` a subclass of `theano.Variable` to end the confusion of what's possible.
But essentially, that's what a layer is right now. Especially with the more specialized layers popping up all over the place. They are theano operations, but with the sanity of keeping shape information and separating function from implementation.

Also, I would add -, /, ** too... but I think their use-case is rather limited.

> @317070 Would operators really change that much?
Well, the main argument I read here, is that there is worry on creating confusion between Theano, Numpy and Lasagne objects. Retrospectively, I think the exact reason I'd like these operations to be there, is because I already am confused on a regular basis. I often need to go back in the code because I treated lasagna layers as Theano tensors. Or needed to create new layers for trivial tasks:
https://github.com/317070/kaggle-heart/blob/master/deep_learning_layers.py#L268
https://github.com/317070/kaggle-heart/blob/master/layers.py#L77
https://github.com/317070/kaggle-heart/blob/master/layers.py#L90
https://github.com/317070/kaggle-heart/blob/master/layers.py#L350
(although, this is code from 8 months ago. Some of them might already be easier to do)

It makes the code more verbose and harder to read. I always find Polish notation difficult to process. I still have some np.dot() nightmare code laying around.

So for me, adding these operators would resolve some of the confusion by effectively doing what I expected it to do on the point of confusion. So, I would argue the confusion is already there, and it might be better to mitigate the problem, rather than fight it.

> @317070 Would operators really change that much?
Well, no. They are after all operators which will do something you can already do. It's that I regularly catch myself trying to use these operators, instead of the more verbose current approach. I think that's an indication it might be beneficial to allow the use of these operators instead.

**Comment by f0k** [comment]
> They are theano operations, but with the sanity of keeping shape information and separating function from implementation.

True. Possibly the only reason holding us back from allowing things such as `T.dot(layer1, layer2)` is that Theano doesn't have shape inference, and it seemed easier to write Lasagne than to fix this.

> I often need to go back in the code because I treated lasagna layers as Theano tensors.

Hmm, interesting. But is this limited to `+` and `*`, or all kind of operations?

> https://github.com/317070/kaggle-heart/blob/master/deep_learning_layers.py#L268

Okay, this would currently be `lasagne.layers.ScaleLayer(layer, theano.constant(3))` and could be `3 * layer`.

> https://github.com/317070/kaggle-heart/blob/master/layers.py#L77

That's a bit more complicated, you're broadcasting by padding the shape on the right rather than the left (as standard broadcasting rules would do). We wouldn't want `*` to break conventions of numpy and Theano. How would you go about this? Implement `__getitem__` with support for `np.newaxis`? Implement `Layer.shape_padright()`?

> https://github.com/317070/kaggle-heart/blob/master/layers.py#L90

Okay, this would require `__rsub__` or writing `1 + -1*layer` instead of `1 - layer`. But if you wrote `1 - layer`, what would this become in terms of Lasagne layers?

I'm still not convinced either way. I like syntactic sugar, but I fear it might open a can of [syntactic sugar worms](https://www.skinnymom.com/wp-content/uploads/2015/05/gummy-worms-sugar-copy-650x432.jpg).
